p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,341 @@
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+func (c *connection) LocalPrivateKey() ic.PrivKey {
+	return c.privKey
+}
+
+func (c *connection) RemotePeer() peer.ID {
+	return c.remotePeer
+}
+
+func (c *connection) RemotePublicKey() ic.PubKey {
+	return c.remoteKey
+}
+
+// implement network.ConnMultiaddrs
+func (c *connection) LocalMultiaddr() ma.Multiaddr {
+	return c.localMultiaddr
+}
+
+func (c *connection) RemoteMultiaddr() ma.Multiaddr {
+	return c.remoteMultiaddr
+}
+
+// implement network.ConnScoper
+func (c *connection) Scope() network.ConnScope {
+	return c.scope
+}
+
+func (c *connection) Transport() tpt.Transport {
+	return c.transport
+}
+
+func (c *connection) addStream(stream *webRTCStream) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.streams[stream.id] = stream
+}
+
+func (c *connection) removeStream(id uint16) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	delete(c.streams, id)
+}
+
+func (c *connection) onConnectionStateChange(state webrtc.PeerConnectionState) {
+	log.Debugf("[%s][%d] handling peerconnection state: %s", c.localPeer, c.dbgId, state)
+	if state == webrtc.PeerConnectionStateFailed || state == webrtc.PeerConnectionStateClosed {
+		// reset any streams
+		c.resetStreams()
+		c.cancel()
+		// is done safe to call twice?
+		c.scope.Done()
+		c.pc.Close()
+	}
+}
+
+// detachChannel detaches an outgoing channel by taking into account the context
+// passed to `OpenStream` as well the closure of the underlying peerconnection
+func (c *connection) detachChannel(ctx context.Context, dc *webrtc.DataChannel) (rwc datachannel.ReadWriteCloser, err error) {
+	done := make(chan struct{})
+	var cls sync.Once
+	defer cls.Do(func() {
+		close(done)
+	})
+	// OnOpen will return immediately for detached datachannels
+	// refer: https://github.com/pion/webrtc/blob/7ab3174640b3ce15abebc2516a2ca3939b5f105f/datachannel.go#L278-L282
+	dc.OnOpen(func() {
+		rwc, err = dc.Detach()
+		// this is safe since the function should return instantly if the peerconnection is closed
+		cls.Do(func() { close(done) })
+	})
+	select {
+	case <-c.ctx.Done():
+		return nil, fmt.Errorf("connection closed")
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-done:
+	}
+	return
+}
+
+// only used during connection setup
+func (c *connection) setRemotePeer(id peer.ID) {
+	c.remotePeer = id
+}
+
+// only used during connection setup
+func (c *connection) setRemotePublicKey(key ic.PubKey) {
+	c.remoteKey = key
+}
+
+// sidAllocator is a helper struct to allocate stream IDs for datachannels. ID
+// reuse is not currently implemented. This prevents streams in pion from hanging
+// with `invalid DCEP message` errors.
+// The id is picked using the scheme described in:
+// https://datatracker.ietf.org/doc/html/draft-ietf-rtcweb-data-channel-08#section-6.5
+// By definition, the DTLS role for inbound connections is set to DTLS Server,
+// and outbound connections are DTLS Client.
+type sidAllocator struct {
```

Comment:
A TODO I had here was to add stream ID reuse. We can know when an ID is available for reuse when the `removeStream` method is called.

Reply:
Ok I added a TODO, will tackle it as part of other TODO's in coming days/weeks.